specification/under_development/profiles/Adaptive_Profile.md

```diff
@@ -0,0 +1,988 @@
+quantum programs.
+
+Here is an example program that mixes mid-circuit measurement (**Bullet 5**), a `qis` function to 
+convert a measurement result to a boolean (**Bullet 7**), and a branch instruction (**Bullet 4**) to control
+how gates are applied based on mid-circuit measurement.
+```llvm
+  tail call void @__quantum__qis__h__body(%Qubit* null)
+  tail call void @__quantum__qis__mz__body(%Qubit* null, %Result* null)
+  tail call void @__quantum__qis__reset__body(%Qubit* null)
+  %0 = tail call i1 @__quantum__rt__read_result__body(%Result* null)
+  br i1 %0, label %then, label %continue
+then:                                   ; preds = %entry
+  tail call void @__quantum__qis__x__body(%Qubit* nonnull inttoptr (i64 1 to %Qubit*))
+  br label %continue__1.i.i.i
+
+continue:
+...
+```
+
+**Bullet 7: Measurement Results to Booleans** <br />
+A function to turn measurement results into classical data, mid-circuit, is critical to using
+forward branching and conditional quantum execution.
+
+```llvm
+%0 = tail call i1 @__quantum__rt__read_result__body(%Result* null)
+  br i1 %0, label %then, label %continue
+```
+
+## Optional Capabilities
+
+**Bullet 8: Qubit resetting** <br />
+If a back-end opts into supporting this feature, then we can allow
+for an instruction to reset qubits back to the |0> state to be defined
+in the instruction set. This can be illustrated by the following program:
+
+```llvm
+call quantum__qis__h__body(%Qubit* null) ; arbitrary gate here
+...
+call quantum__qis__reset__body(%Qubit* null)
+call quantum__qis__h__body(%Qubit* nonnull inttoptr (i64 1 to %Qubit*)) ; arbitrary gate choice again
+...
+```
+
+This can be combined with the mid-circuit measurement (**Bullet 5**) to allow for the capability to re-use
+the same qubit for computations, limiting the amount of qubits required
+for redundant calculations and also allowing for circuit width to bet reduced in favor of increasing depth:
+
+```llvm
+call quantum__qis__h__body(%Qubit* null) 
+call quantum__qis__mz__body(%Qubit* null, %Result* null) ; 1st coin flip
+call quantum__qis__reset__body(%Qubit* null)
+call quantum__qis__h__body(%Qubit* null) 
+call quantum__qis__mz__body(%Qubit* null, %Result* null) ; 2nd coin flip
+call quantum__qis__reset__body(%Qubit* null)
+```
+
+**Bullet 9: Classical integer computations** <br />
+An Adaptive Profile program may include expressions for numeric and logical computations
+that don't involve allocating memory for aggregate data structures. These
+can include integer arithmetic calcuations of a back-end specified width, floating
+point arithmetic calculations of a back-end specified width, or logical operations
+on integers of a back-end specified width. Front-ends can use any integer width when generating code,
+and it is the responsiblity of a back-end to provide a compile-time error message if the front-end
+uses an integer width greater than that which the support.
```

Comment:
two concerns:
* an end-user would be working with a "high level" data type which is likely to imply high width (so how would they know), and
* therefore some mechanism for lowering the width would need to happen somewhere (breaking this guidance)

Personally, I wouldn't be excited about having to deal with explicit typing (e.g., numpy's `dtype`) either as an end-user or working on a backend. On the other hand, rigor is good too...

Reply:
We intend to update the spec to have the module flag explicitly specifying the used data type size(s). The backend should then look at these flags and reject the program if it does not provide support for these sizes.